### DISCUSSION TITLE
Validating URLs without scheme

### DISCUSSION THREAD

**Discussion by mblakesley** [pydantic/pydantic#3014] [Question] [answered]
I'm looking for the ability to validate URLs that don't have a scheme in front, e.g., `example.com` all by itself, without `http://` in front.

I was hoping to use `AnyHttpUrl` out of the box:
```
class Hostname(BaseModel):
    domain: AnyHttpUrl

Hostname(domain="example.com")  # fails, even though we all know that's a fine domain name
```
I can't find a good way around this. The best I can come up with involves using `validator`s to actually modify the domain name specified by the user, prepending `http://`, letting validation happen, then reverting that change. That seems like kind of a bad solution.

Can anyone help me with a better solution?

Am I really the first person to want this?

**Comment by lesnik512** [comment]
Maybe this will help

```python
from pydantic import BaseModel, ConstrainedInt, ConstrainedStr
import re

class UrlWithoutSchema(ConstrainedStr):
    min_length = 1
    max_length = 2000
    strip_whitespace = True
    regex = re.compile(r'^[A-Za-z0-9][A-Za-z0-9.-]+(:\d+)?(/.*)?$')

class Hostname(BaseModel):
    domain: UrlWithoutSchema

Hostname(domain="example.com/fsfa/fsf")
```

**Reply by mblakesley** [reply]
Thank you for the attempt, but yeah, I'd rather not roll my own validation. There's a whole pre-written validation library here with Pydantic. That's its purpose! I'm looking for a way to tap into its goodness.

**Comment by ryneeverett** [comment, marked as answer]
Another workaround would be to subclass `AnyUrl` and remove the scheme validation:

```py
class Url(pydantic.AnyUrl):
    """ Override to allow scheme-less url's. """

    @classmethod
    def validate_parts(cls, parts: Dict[str, str]) -> Dict[str, str]:
        """
        A method used to validate parts of an URL.
        Could be overridden to set default values for parts if missing
        """
        # scheme = parts['scheme']
        # if scheme is None:
            # raise errors.UrlSchemeError()

        # if cls.allowed_schemes and scheme.lower() not in cls.allowed_schemes:
            # raise errors.UrlSchemePermittedError(cls.allowed_schemes)

        port = parts['port']
        if port is not None and int(port) > 65_535:
            raise errors.UrlPortError()

        user = parts['user']
        if cls.user_required and user is None:
            raise errors.UrlUserInfoError()

        return parts

```

I'm curious though if the ability to validate schema-less urls was intentionally omitted or just hasn't been implemented yet.

**Reply by mblakesley** [reply]
Good idea! Thank you!

**Reply by thinktrans** [reply]
Just a heads up while chalking out the details of "A method used to validate parts of an URL.". Acknowledging the presence of the Internationalized Email IDs would be highly desirable. 
A synchronized use of the email_validator, idna and smtplib from Python 3 would definitely ensure that.
